api/handlers/security_group_test.go

```diff
@@ -0,0 +1,178 @@
+				DisplayName: "test-security-group",
+				Rules: []payloads.SecurityGroupRule{
+					{
+						Protocol:    korifiv1alpha1.ProtocolTCP,
+						Ports:       "80",
+						Destination: "192.168.1.1",
+					},
+				},
+				Relationships: payloads.SecurityGroupRelationships{
+					RunningSpaces: payloads.ToManyRelationship{
+						Data: []payloads.RelationshipData{
+							{
+								GUID: "space1",
+							},
+						},
+					},
+					StagingSpaces: payloads.ToManyRelationship{
+						Data: []payloads.RelationshipData{
+							{
+								GUID: "space2",
+							},
+						},
+					},
+				},
+			}
+			requestValidator.DecodeAndValidateJSONPayloadStub = decodeAndValidatePayloadStub(&payload)
+
+			securityGroupRepo.CreateSecurityGroupReturns(repositories.SecurityGroupRecord{
+				GUID: "test-guid",
+				Name: "test-security-group",
+				Rules: []repositories.SecurityGroupRule{
+					{
+						Protocol:    korifiv1alpha1.ProtocolTCP,
+						Ports:       "80",
+						Destination: "192.168.1.1",
+					},
+				},
+				RunningSpaces: []string{"space1"},
+				StagingSpaces: []string{"space2"},
+			}, nil)
+
+			spaceRepo.ListSpacesReturns([]repositories.SpaceRecord{
+				{
+					Name: "space1",
+				},
+				{
+					Name: "space2",
+				},
+			}, nil)
+		})
+
+		It("validates the request", func() {
+			Expect(requestValidator.DecodeAndValidateJSONPayloadCallCount()).To(Equal(1))
+			actualReq, _ := requestValidator.DecodeAndValidateJSONPayloadArgsForCall(0)
+			Expect(bodyString(actualReq)).To(Equal("the-json-body"))
+		})
+
+		It("creates a security group with a rule", func() {
+			Expect(securityGroupRepo.CreateSecurityGroupCallCount()).To(Equal(1))
+			Expect(spaceRepo.ListSpacesCallCount()).To(Equal(1))
+
+			_, actualAuthInfo, createSecurityGroupMessage := securityGroupRepo.CreateSecurityGroupArgsForCall(0)
+			Expect(actualAuthInfo).To(Equal(authInfo))
+			Expect(createSecurityGroupMessage.DisplayName).To(Equal("test-security-group"))
+			Expect(createSecurityGroupMessage.Rules).To(Equal([]repositories.SecurityGroupRule{
+				{
+					Protocol:    korifiv1alpha1.ProtocolTCP,
+					Ports:       "80",
+					Destination: "192.168.1.1",
+				},
+			}))
+
+			_, _, listSpacesMessage := spaceRepo.ListSpacesArgsForCall(0)
+			Expect(listSpacesMessage.GUIDs).To(ConsistOf("space1", "space2"))
+
+			Expect(rr).To(HaveHTTPStatus(http.StatusCreated))
+			Expect(rr).To(HaveHTTPHeaderWithValue("Content-Type", "application/json"))
+			Expect(rr).To(HaveHTTPBody(SatisfyAll(
+				MatchJSONPath("$.guid", "test-guid"),
+				MatchJSONPath("$.name", "test-security-group"),
+				MatchJSONPath("$.rules[0].protocol", "tcp"),
+				MatchJSONPath("$.rules[0].ports", "80"),
+				MatchJSONPath("$.rules[0].destination", "192.168.1.1"),
+				MatchJSONPath("$.relationships.running_spaces.data[0].guid", "space1"),
+				MatchJSONPath("$.relationships.staging_spaces.data[0].guid", "space2"),
+			)))
+		})
+
+		When("a space in the security group does not exist", func() {
```

Comment:
We should have another context on this level to test that the handler returns an error when listing spaces fails. The space repo could be stubbed to return a random error and the test should expect an unknown error. No need to check for various error types, as the handler just returns the error from the repository